CPA LockoutPanel UI

wilkie opened the pull request:
This adds `LockoutPanel` and the associated route, images, haml page, and JS initialization in our pipeline.

There is no behavior, just the UI, as per our ticket. Gluing this UI and the API/data model together is a next step.

But I did go ahead and add a route for the page for convenience.

Added each respective thingy to the `sessions` namespace... not sure if that's a good place. Felt right and better than creating something new or have it float around with everything else.

### Behaviors of the UI:

* Not currently masked behind the CPA experiment flag... not necessary: not in any common path or updating any already used UI.
* `/lockout` route is a placeholder... any failed authentication on parental permission should just redirect there.
* Information is passed by adding data attributes on the containing div currently. Easy to change.
* Submit button is disabled until a valid email is typed in.
* Resend button (styled as a link) replaces the email with the pending email and submits.
* Delete by date is rendered as UTC so there are no strange timezone shenanigans when rendering it. Let's be sure to generate a UTC delete by date.
* Last request date is rendered in the current locale / timezone. We should still store it as UTC.

### New Strings for i18n:

`sessionLockoutNewAccountHeader`: "Just one more thing!"
`sessionLockoutPendingHeader`: "Thanks! We've contacted your parent."
`sessionLockoutPrompt`: "We need your parent to approve your account before you can get started. Please supply us with your parent’s email address so they can grant you permission."
`sessionLockoutPendingPrompt`: "We sent an email to {pendingEmail}. Didn’t receive anything? Update your parent’s email below or send another request."
`sessionLockoutNote`: "Note: Your account will be deleted if we do not receive your parent’s permission by {deleteDate}."
`sessionLockoutParentEmailField`: "Parent Email:"
`sessionLockoutParentStatusField`: "Parent Permission:"
`sessionLockoutSubmit`: "Send permission request"
`sessionLockoutStatusPending`: "Pending"
`sessionLockoutStatusNotSubmitted`: "Not Submitted"
`sessionLockoutResendEmail`: "Resend permission email"
`sessionLockoutUpdateSubmit`: "Update and send"
`signOutButton`: "Sign out",

## Links

- jira ticket: [P20-61](https://codedotorg.atlassian.net/browse/P20-61)

## Testing story

Manual testing of the basic boilorplate behaviors. Then, contributing one Storybook with two templates:

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

### New Accounts

![image](https://github.com/code-dot-org/code-dot-org/assets/31674/d2e2ce4a-5861-4944-9ba0-5daac392e1e0)

### When there is a pending email request:

![image](https://github.com/code-dot-org/code-dot-org/assets/31674/6d4a1bce-fc14-47e1-82f9-232c650f37b1)

### New Accounts (right-to-left)

![image](https://github.com/code-dot-org/code-dot-org/assets/31674/85f3d1ac-9303-4944-afb4-c037bcf6d711)

### Pending email (right-to-left)

![image](https://github.com/code-dot-org/code-dot-org/assets/31674/2881e7c6-bca9-4f82-8242-cb59995068ea)

## Deployment strategy

No experiment flag. We can just deploy the UI without the glue since nobody will know to look at it yet. :)

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

wilkie commented:
Amended quickly to fix a commented out line... I was debugging. Probably got that in before anybody saw. :)

carl-codeorg commented on apps/i18n/common/en_us.json:
```diff
@@ -1946,6 +1946,18 @@
   "seeIncubatorProjects": "See Incubator projects",
+  "sessionLockoutNewAccountHeader": "Just one more thing!",
+  "sessionLockoutPendingHeader": "Thanks! We've contacted your parent.",
+  "sessionLockoutPrompt": "We need your parent to approve your account before you can get started. Please supply us with your parent’s email address so they can grant you permission.",
```
Is a provision in the law for a legal guardian (but not a parent) to provide permission? If so do we need language like "parent or guardian"?

wilkie replied:
That is a good question. I noticed this language already used on our existing parent email process, so it seems like a good idea to ask to see if we should just conform to that language.

wilkie replied:
We are updating the text! Pushing that up with the next commit.

elf-code commented on apps/i18n/common/en_us.json:
```diff
@@ -2114,6 +2126,7 @@
   "showWords": "Show words",
   "showingLibrariesFromSection": "Showing libraries from section: ",
   "signup": "Sign up for the intro course",
+  "signOutButton": "Sign out",
```
Is there an existing string for this? (I see it in English and Spanish so there's something that gets translated that means Sign Out)

daynew replied:
I don't think there is a "Sign Out" in this file already.

wilkie replied:
No, there is no established string for "Sign out" available to the front-end. 🥲

elf-code commented on apps/src/templates/sessions/LockoutPanel.jsx:
```diff
@@ -0,0 +1,283 @@
+ * permission. This panel gives a form to request (or see details about a
+ * pending request for) parental permission.
+ */
+class LockoutPanel extends Component {
```
Can we make a new React component a functional component instead of a class component? I'm happy to pair with you to shift around the syntax or twist it around myself if that's easier.

wilkie replied:
Do you have an example of a functional component that is similar to our case here somewhere in our codebase to use to conform to convention?

wilkie replied:
I'd particularly like an example of a component with form interactivity... that's the challenge I see in re-wrapping it in the functional style.

elf-code replied:
Here's an example in our codebase (I think it has a lot of interactions) -- https://github.com/code-dot-org/code-dot-org/blob/staging/apps/src/music/views/ChordPanel.tsx

Hooks are weird (I think useState is the only semi-intuitive one, but the example component has a couple different hooks in it). But the React.dev notes on hooks are good - https://react.dev/reference/react

wilkie replied:
This is helpful! Good to see the conventions to follow... I already did the work last night... pretty straightforward. Much prefer the "functional" style.

wilkie replied:
Done. Ish. Do I need a `useCallback` if there's no state being used? It's just reacting to input events. It's always the same function.

Global redux state can't be wrapped very effectively, I don't think... but the things being pulled from that are the `isRTL` bool which can only be changed via changing the locale which is always a page refresh anyway.

elf-code replied:
I don't think you need a `useCallback` and yes, I agree on the redux store and `isRTL`

elf-code reviewed (commented):
Is there a way to directly view the Storybook as part of testing? Can you include the path/screenshot to that (for learning purposes)?

daynew commented on apps/src/templates/sessions/LockoutPanel.jsx:
```diff
@@ -0,0 +1,283 @@
+    padding: 6,
+  },
+  fieldSection: {
+    display: 'table-cell',
```
My understanding is that it's better to use [Flexbox](https://css-tricks.com/snippets/css/a-guide-to-flexbox/) for layout stuff instead of trying to use tables or table styling. Did you already try using it?

wilkie replied:
That's kinda true. I could try to use a full width flex box for that section and grow the field side. I think if I establish the flex basis, it won't squish the text. Hmm. Yes, I think that would do the same thing.

wilkie replied:
Gosh expanding `<input>` tags to the full width of grown divs in a flexbox is silly annoying, but I have done it... will test it a little and push it up.

wilkie replied:
Done.

wilkie commented:
> Is there a way to directly view the Storybook as part of testing? Can you include the path/screenshot to that (for learning purposes)?

The screenshots of the storybooks are more or less how they are depicted in the PR description. Our storybook view does not have root application-level styling, so the headers look wrong (as they do on all other storybook components), but otherwise look the same as the site.

You view the storybooks by pulling down my branch and then running the storybook web server and visiting it (localhost, port 9001, I think):

```
yarn storybook
```

I do not believe we have a staging storybook server set up at all? That'd be nice.

![image](https://github.com/code-dot-org/code-dot-org/assets/31674/280228f8-8cc5-44e4-8eff-81be821644bd)

(Ignore the broken image... it works... I'm just too lazy to restart my web server)

![image](https://github.com/code-dot-org/code-dot-org/assets/31674/5b572c1a-8ef3-40c5-95b8-79e52f45ad29)

wilkie commented:
I added screenshots of the right-to-left rendering of the widget. We do not have a convention for creating storybooks for right-to-left... nor do we have a strategy of rendering components right-to-left in the storybook view. Hmm.

elf-code reviewed (approved):
LGTM! Also I'm assuming you double-checked that the functional changes don't change the visuals.

wilkie commented:
Indeed. In fact, the PR description has been updated with new visual screens which include RTL rendering. I'll verify that the storybooks render later today before merging, once I get Dayne's approval on their requested changes.

daynew commented on apps/src/templates/sessions/LockoutPanel.jsx:
```diff
@@ -0,0 +1,280 @@
+  return (
+    <div style={styles.container} className="lockout-panel">
+      {/* Header image: Depends of if permission request is sent. */}
+      <img
```
alt text for both images

wilkie replied:
I've added them and the strings will be localized.

daynew commented on apps/src/templates/sessions/LockoutPanel.jsx:
```diff
@@ -0,0 +1,280 @@
+    day: 'numeric',
+  };
+
+  // We want to mark the email in the prompt in bold, so we split the
+  // localized string. It is a slight hack since we cannot just embed React
+  // component code into a string. Should work as though we could though.
+  const pendingPrompt = i18n.sessionLockoutPendingPrompt({
```
I wonder if the `EnhancedSafeMarkdown` component would work? I think that is how we render markdown strings in other React components. I think your solution works as well though.

wilkie replied:
I'm gonna leave it as is, then. I don't think it is worth potentially overcomplicating the string and worry about what happens if markdown ends up in the translated strings. Pity there isn't a good way of just doing string interpolation with localized strings and injecting simple markdown or html without needlessly going with markdown 100%.

daynew commented on dashboard/config/routes.rb:
```diff
@@ -197,6 +197,7 @@
       get '/users/demigrate_from_multi_auth', to: 'registrations#demigrate_from_multi_auth'
       get '/users/to_destroy', to: 'registrations#users_to_destroy'
       get '/reset_session', to: 'sessions#reset'
+      get '/lockout', to: 'sessions#lockout'
```
What if a student visits this URL and they have permission? I'm thinking we should redirect them to the /home page, but that work should be done in the global redirect task. I will note it there.

daynew commented on apps/src/templates/sessions/LockoutPanel.jsx:
```diff
@@ -0,0 +1,280 @@
+        <div style={styles.statusSection}>
+          <label
+            style={isRTL ? styles.statusLabelRTL : styles.statusLabel}
+            htmlFor="permission-status"
```
Oh cool, I didn't know you had to use `htmlFor` instead of `for` in JSX. That makes sense. Thanks for looking this up :)

daynew commented on apps/src/templates/sessions/LockoutPanel.jsx:
```diff
@@ -0,0 +1,280 @@
+        {/* The submit button. */}
+        {/* An empty onClick will still submit the form. */}
+        <Button
+          style={{...styles.button, float: isRTL ? 'left' : 'right'}}
```
Maybe a nit, but should we use flexbox to avoid using `float` and `isRTL` customization? Would the flexbox `justify-content: space-between` achieve the same spacing between the "Sign out" and "Send permission request" buttons?

wilkie replied:
This is an amazing idea.

wilkie replied:
That worked really well and now I want to revisit the past 10 years of my life and find all the places where I should have done this. Thank you!

daynew reviewed (approved):
Looks great! Just some minor comments.

wilkie commented:
Will be addressing Dayne's comments and adding the new text for Parent/Guardian for the next pass.

wilkie commented:
Ok. Added the new "parent / guardian" strings. I think it covers all of the cases. Removed all "float"ed content and replaced with some cool flexbox ideas that will work in both LTR and RTL circumstances. I have looked at the visuals and confirm that they still match the ones in the description. No visual changes. Alt text applied to both header images to describe the little penguin :)

artem-vavilov commented on dashboard/app/views/sessions/lockout.html.haml:
```diff
@@ -0,0 +1,4 @@
+#lockout-container{:"data-request-date" => (DateTime.now - 6).iso8601, :"data-delete-date" => (DateTime.now + 2).iso8601, :"data-pending-email" => "", :"data-api-url" => "/permissions"}
```
Just a small recommendation
```suggestion
#lockout-container{data: {request_date: 6.days.ago.iso8601, delete_date: 2.days.since.iso8601, pending_email: '', api_url: '/permissions'}}
```
